File: src/main/java/leetcode/lists/hot100/LeetCode238.java
package leetcode.lists.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/19 13:46
 */
public class LeetCode238 {


    public int[] productExceptSelf(int[] nums) {

        if (nums == null || nums.length == 0) {
            return nums;
        }

        //[1,2,3,4]
        // 1 1 2 6

        //[-1,1,0,-3,3]
        // -1 -1 -1
        int[] one = new int[nums.length];
        one[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            one[i] = nums[i - 1] * one[i - 1];
        }
        int temp = nums[nums.length - 1];
        for (int i = nums.length - 2; i >= 0; i--) {
            one[i] = one[i] * temp;
            temp = temp * nums[i];
        }
        return one;
    }
}

File: go/src/leetcode/hot100/LeetCode238.go
package hot100

func productExceptSelf(nums []int) []int {
	arr := make([]int, len(nums))
	res := make([]int, len(nums))
	arr[len(arr)-1] = 1
	for i := len(nums) - 2; i >= 0; i-- {
		arr[i] = arr[i+1] * nums[i+1]
	}
	res[0] = arr[0]
	pre := 1
	for i := 1; i < len(nums); i++ {
		pre = pre * nums[i-1]
		res[i] = pre * arr[i]
	}
	return res
}
